src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+            // instead the shape of the tree looks like this:
+            //          +
+            //         / \
+            //        +   4
+            //       / \
+            //      +   3
+            //     / \
+            //    1   2
+            // in this case there is no such one node that covers ends of selection and is located inside the selection
+            // to handle this we check if both start and end of the selection belong to some binary operation
+            // and start node is parented by the parent of the end node
+            // if this is the case - expand the selection to the entire parent of end node (in this case it will be [1 + 2 + 3] + 4)
+            const startParent = skipParentheses(start.parent);
+            const endParent = skipParentheses(end.parent);
+            if (isBinaryExpression(startParent) && isBinaryExpression(endParent) && isNodeDescendantOf(startParent, endParent)) {
+                start = end = endParent;
+            }
+            else {
+                // start and end nodes belong to different subtrees
+                return createErrorResult(sourceFile, span.start, length, Messages.CannotExtractFunction);
+            }
+        }
+        if (start !== end) {
+            // start and end should be statements and parent should be either block or a source file
+            if (!isBlockLike(start.parent)) {
+                return createErrorResult(sourceFile, span.start, length, Messages.CannotExtractFunction);
+            }
+            const statements: Statement[] = [];
+            for (const statement of (<BlockLike>start.parent).statements) {
+                if (statement === start || statements.length) {
+                    const errors = checkNode(statement);
+                    if (errors) {
+                        return { errors };
+                    }
+                    statements.push(statement);
+                }
+                if (statement === end) {
+                    break;
+                }
+            }
+            return { targetRange: { range: statements, facts: rangeFacts, declarations } };
+        }
+        else {
+            // We have a single expression (start)
+            const errors = checkRootNode(start) || checkNode(start);
+            if (errors) {
+                return { errors };
+            }
+
+            // If our selection is the expression in an ExrpessionStatement, expand
+            // the selection to include the enclosing Statement (this stops us
+            // from trying to care about the return value of the extracted function
+            // and eliminates double semicolon insertion in certain scenarios)
+            const range = isStatement(start)
+                ? [start]
+                : start.parent && start.parent.kind === SyntaxKind.ExpressionStatement
+                    ? [start.parent as Statement]
+                    : <Expression>start;

```

Comment:
><Expression>start [](start = 22, length = 17)

Is this cast guaranteed to succeed?  There are no other types that `start` could have?